File: python/src/binary_market_data/discovery.py
"""Public catalog pages and condition-to-token resolution, with explicit pagination."""

from dataclasses import dataclass, replace
from urllib.parse import quote

import httpx

from ._http import client_context, endpoint, get_text
from .types import JsonRecord, Venue
from .venue._parsing import ParseError, array, load, obj

POLYMARKET_CATALOG = "https://gamma-api.polymarket.com"
POLYMARKET_CLOB = "https://clob.polymarket.com"
KALSHI_REST = "https://external-api.kalshi.com/trade-api/v2"
KALSHI_STATUSES = ("open", "unopened", "paused", "closed", "settled")


class DiscoveryError(ValueError):
    def __init__(self, code: str) -> None:
        self.code = code
        super().__init__(code)


@dataclass(frozen=True, slots=True)
class OutcomeInstrument(JsonRecord):
    outcome: str | None
    instrument_id: str


@dataclass(frozen=True, slots=True)
class MarketDescriptor(JsonRecord):
    venue: Venue
    market_id: str
    title: str | None
    slug: str | None
    status: str
    book_enabled: bool
    accepting_orders: bool
    close_time: str | None
    instruments: tuple[OutcomeInstrument, ...]

    def book_instrument_ids(self) -> list[str]:
        return [instrument.instrument_id for instrument in self.instruments]


@dataclass(frozen=True, slots=True)
class MarketPage(JsonRecord):
    venue: Venue
    markets: tuple[MarketDescriptor, ...]
    next_cursor: str | None


@dataclass(frozen=True, slots=True)
class PolymarketDiscoveryConfig:
    limit: int = 20
    catalog_base: str = POLYMARKET_CATALOG
    cursor: str | None = None
    query: str | None = None


@dataclass(frozen=True, slots=True)
class KalshiDiscoveryConfig:
    limit: int = 20
    rest_base: str = KALSHI_REST
    cursor: str | None = None
    query: str | None = None
    status: str | None = "open"


def _text(value: dict, *keys: str) -> str | None:
    return next((value[k] for k in keys if isinstance(value.get(k), str) and value[k]), None)


def _required(value: dict, *keys: str) -> str:
    result = _text(value, *keys)
    if result is None:
        raise ParseError("identifier")
    return result


def _bool(value: dict, *keys: str, default: bool = False) -> bool:
    return next((value[k] for k in keys if type(value.get(k)) is bool), default)


def _strings(value: object) -> list[str]:
    if value is None:
        return []
    values = array(load(value) if isinstance(value, str) else value, "string_array")
    if not all(isinstance(item, str) and item for item in values):
        raise ParseError("string_array.item")
    return values


def parse_polymarket_page(text: str | bytes) -> MarketPage:
    root = obj(load(text))
    markets = []
    for value in array(root.get("markets"), "markets"):
        value = obj(value)
        tokens, outcomes = _strings(value.get("clobTokenIds")), _strings(value.get("outcomes"))
        if tokens and outcomes and len(tokens) != len(outcomes):
            raise ParseError("outcomes")
        instruments = tuple(
            OutcomeInstrument(outcomes[index] if outcomes else None, token)
            for index, token in enumerate(tokens)
        )
        active, closed, archived = (_bool(value, key) for key in ("active", "closed", "archived"))
        status = (
            "archived" if archived else "closed" if closed else "active" if active else "inactive"
        )
        markets.append(
            MarketDescriptor(
                venue=Venue.POLYMARKET,
                market_id=_required(value, "conditionId", "condition_id"),
                title=_text(value, "question", "title"),
                slug=_text(value, "slug", "market_slug"),
                status=status,
                book_enabled=_bool(
                    value, "enableOrderBook", "enable_order_book", default=bool(tokens)
                ),
                accepting_orders=_bool(
                    value, "acceptingOrders", "accepting_orders", default=active and not closed
                ),
                close_time=_text(value, "endDateIso", "endDate", "end_date_iso"),
                instruments=instruments,
            )
        )
    return MarketPage(Venue.POLYMARKET, tuple(markets), _text(root, "next_cursor"))


def parse_kalshi_page(text: str | bytes) -> MarketPage:
    root = obj(load(text))
    markets = []
    for value in array(root.get("markets"), "markets"):
        value = obj(value)
        ticker, status = _required(value, "ticker"), _text(value, "status") or "unknown"
        markets.append(
            MarketDescriptor(
                venue=Venue.KALSHI,
                market_id=ticker,
                title=_text(value, "title", "subtitle", "yes_sub_title"),
                slug=_text(value, "event_ticker"),
                status=status,
                book_enabled=(_text(value, "market_type") or "binary").lower() == "binary",
                accepting_orders=status.lower() in ("open", "active"),
                close_time=_text(
                    value, "close_time", "expiration_time", "expected_expiration_time"
                ),
                instruments=(OutcomeInstrument(None, ticker),),
            )
        )
    return MarketPage(Venue.KALSHI, tuple(markets), _text(root, "cursor"))


def parse_polymarket_market_info(text: str | bytes) -> tuple[OutcomeInstrument, ...]:
    root = obj(load(text))
    tokens = array(root.get("t", root.get("tokens")), "tokens")
    if not tokens:
        raise ParseError("tokens")
    return tuple(
        OutcomeInstrument(_text(obj(token), "o", "outcome"), _required(token, "t", "token_id"))
        for token in tokens
    )


def _filter(page: MarketPage, query: str | None, limit: int) -> MarketPage:
    needle = (query or "").strip().lower()
    return replace(
        page,
        markets=tuple(
            market
            for market in page.markets
            if market.book_enabled
            and (
                not needle
                or any(
                    needle in field.lower()
                    for field in (
                        market.market_id,
                        market.title or "",
                        market.slug or "",
                    )
                )
            )
        )[:limit],
    )


def _limit(limit: int, maximum: int) -> None:
    if type(limit) is not int or not 1 <= limit <= maximum:
        raise DiscoveryError("invalid_catalog_limit")


async def _request(base: str, path: str, params: dict, client: httpx.AsyncClient | None) -> str:
    try:
        url = endpoint(base, path)
    except ValueError:
        raise DiscoveryError("invalid_catalog_endpoint") from None
    try:
        async with client_context(client) as session:
            return await get_text(session, url, params=params)
    except (httpx.HTTPError, ValueError):
        raise DiscoveryError("catalog_request_failed") from None


async def discover_polymarket(
    config: PolymarketDiscoveryConfig | None = None,
    *,
    client: httpx.AsyncClient | None = None,
) -> MarketPage:
    config = config or PolymarketDiscoveryConfig()
    _limit(config.limit, 100)
    params = {"limit": str(config.limit), "closed": "false"}
    if config.cursor is not None:
        params["after_cursor"] = config.cursor
    body = await _request(config.catalog_base, "markets/keyset", params, client)
    try:
        return _filter(parse_polymarket_page(body), config.query, config.limit)
    except ParseError:
        raise DiscoveryError("invalid_catalog_response") from None


async def discover_kalshi(
    config: KalshiDiscoveryConfig | None = None,
    *,
    client: httpx.AsyncClient | None = None,
) -> MarketPage:
    config = config or KalshiDiscoveryConfig()
    _limit(config.limit, 1000)
    if config.status is not None and config.status not in KALSHI_STATUSES:
        raise DiscoveryError("invalid_catalog_status")
    params = {"limit": str(config.limit)}
    if config.cursor is not None:
        params["cursor"] = config.cursor
    if config.status is not None:
        params["status"] = config.status
    body = await _request(config.rest_base, "markets", params, client)
    try:
        return _filter(parse_kalshi_page(body), config.query, config.limit)
    except ParseError:
        raise DiscoveryError("invalid_catalog_response") from None


async def resolve_polymarket_condition(
    clob_base: str,
    condition_id: str,
    *,
    client: httpx.AsyncClient | None = None,
) -> tuple[OutcomeInstrument, ...]:
    if not condition_id.strip() or condition_id in (".", ".."):
        raise DiscoveryError("invalid_condition_id")
    body = await _request(clob_base, "clob-markets/" + quote(condition_id, safe=""), {}, client)
    try:
        return parse_polymarket_market_info(body)
    except ParseError:
        raise DiscoveryError("invalid_catalog_response") from None

File: src/discovery.rs
use crate::types::Venue;
use reqwest::{Client, Url};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::time::Duration;
use thiserror::Error;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct OutcomeInstrument {
    pub outcome: Option<String>,
    pub instrument_id: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MarketDescriptor {
    pub venue: Venue,
    pub market_id: String,
    pub title: Option<String>,
    pub slug: Option<String>,
    pub status: String,
    pub book_enabled: bool,
    pub accepting_orders: bool,
    pub close_time: Option<String>,
    pub instruments: Vec<OutcomeInstrument>,
}

impl MarketDescriptor {
    pub fn book_instrument_ids(&self) -> Vec<String> {
        self.instruments
            .iter()
            .map(|instrument| instrument.instrument_id.clone())
            .collect()
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct MarketPage {
    pub venue: Venue,
    pub markets: Vec<MarketDescriptor>,
    pub next_cursor: Option<String>,
}

#[derive(Debug, Clone)]
pub struct PolymarketDiscoveryConfig {
    pub catalog_base: String,
    pub limit: usize,
    pub cursor: Option<String>,
    pub query: Option<String>,
}

impl PolymarketDiscoveryConfig {
    pub fn open_markets(limit: usize) -> Self {
        Self {
            catalog_base: "https://gamma-api.polymarket.com".to_string(),
            limit,
            cursor: None,
            query: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct KalshiDiscoveryConfig {
    pub rest_base: String,
    pub limit: usize,
    pub cursor: Option<String>,
    pub query: Option<String>,
    pub status: Option<String>,
}

impl KalshiDiscoveryConfig {
    pub fn open_markets(limit: usize) -> Self {
        Self {
            rest_base: "https://external-api.kalshi.com/trade-api/v2".to_string(),
            limit,
            cursor: None,
            query: None,
            status: Some("open".to_string()),
        }
    }
}

#[derive(Debug, Error)]
pub enum DiscoveryError {
    #[error("catalog endpoint is invalid")]
    InvalidEndpoint,
    #[error("catalog limit is outside the supported range")]
    InvalidLimit,
    #[error("catalog request failed")]
    Request(#[source] reqwest::Error),
    #[error("catalog response is invalid")]
    Json(#[source] serde_json::Error),
    #[error("catalog response is missing required field: {0}")]
    Field(&'static str),
}

pub async fn discover_polymarket(
    config: &PolymarketDiscoveryConfig,
) -> Result<MarketPage, DiscoveryError> {
    if !(1..=100).contains(&config.limit) {
        return Err(DiscoveryError::InvalidLimit);
    }
    let mut url = endpoint(&config.catalog_base, "markets/keyset")?;
    {
        let mut query = url.query_pairs_mut();
        query.append_pair("limit", &config.limit.to_string());
        query.append_pair("closed", "false");
        if let Some(cursor) = &config.cursor {
            query.append_pair("after_cursor", cursor);
        }
    }
    let body = request_text(url).await?;
    let mut page = parse_polymarket_page(&body)?;
    filter_page(&mut page, config.query.as_deref(), config.limit);
    Ok(page)
}

pub async fn discover_kalshi(config: &KalshiDiscoveryConfig) -> Result<MarketPage, DiscoveryError> {
    if !(1..=1000).contains(&config.limit) {
        return Err(DiscoveryError::InvalidLimit);
    }
    let mut url = endpoint(&config.rest_base, "markets")?;
    {
        let mut query = url.query_pairs_mut();
        query.append_pair("limit", &config.limit.to_string());
        if let Some(cursor) = &config.cursor {
            query.append_pair("cursor", cursor);
        }
        if let Some(status) = &config.status {
            query.append_pair("status", status);
        }
    }
    let body = request_text(url).await?;
    let mut page = parse_kalshi_page(&body)?;
    filter_page(&mut page, config.query.as_deref(), config.limit);
    Ok(page)
}

pub async fn resolve_polymarket_condition(
    clob_base: &str,
    condition_id: &str,
) -> Result<Vec<OutcomeInstrument>, DiscoveryError> {
    if condition_id.trim().is_empty() {
        return Err(DiscoveryError::Field("condition_id"));
    }
    let mut url = endpoint(clob_base, "clob-markets")?;
    url.path_segments_mut()
        .map_err(|_| DiscoveryError::InvalidEndpoint)?
        .push(condition_id);
    let body = request_text(url).await?;
    parse_polymarket_market_info(&body)
}

pub fn parse_polymarket_page(text: &str) -> Result<MarketPage, DiscoveryError> {
    let root: Value = serde_json::from_str(text).map_err(DiscoveryError::Json)?;
    let values = root
        .get("markets")
        .and_then(Value::as_array)
        .ok_or(DiscoveryError::Field("markets"))?;
    let mut markets = Vec::with_capacity(values.len());
    for value in values {
        let market_id = required_string(value, &["conditionId", "condition_id"])?;
        let token_ids = string_list(value.get("clobTokenIds"))?;
        let outcomes = string_list(value.get("outcomes"))?;
        let instruments = pair_instruments(token_ids, outcomes)?;
        let active = optional_bool(value, &["active"]).unwrap_or(false);
        let closed = optional_bool(value, &["closed"]).unwrap_or(false);
        let archived = optional_bool(value, &["archived"]).unwrap_or(false);
        let book_enabled = optional_bool(value, &["enableOrderBook", "enable_order_book"])
            .unwrap_or(!instruments.is_empty());
        markets.push(MarketDescriptor {
            venue: Venue::Polymarket,
            market_id,
            title: optional_string(value, &["question", "title"]),
            slug: optional_string(value, &["slug", "market_slug"]),
            status: market_status(active, closed, archived),
            book_enabled,
            accepting_orders: optional_bool(value, &["acceptingOrders", "accepting_orders"])
                .unwrap_or(active && !closed),
            close_time: optional_string(value, &["endDateIso", "endDate", "end_date_iso"]),
            instruments,
        });
    }
    Ok(MarketPage {
        venue: Venue::Polymarket,
        markets,
        next_cursor: optional_string(&root, &["next_cursor"]),
    })
}

pub fn parse_kalshi_page(text: &str) -> Result<MarketPage, DiscoveryError> {
    let root: Value = serde_json::from_str(text).map_err(DiscoveryError::Json)?;
    let values = root
        .get("markets")
        .and_then(Value::as_array)
        .ok_or(DiscoveryError::Field("markets"))?;
    let mut markets = Vec::with_capacity(values.len());
    for value in values {
        let ticker = required_string(value, &["ticker"])?;
        let status = optional_string(value, &["status"]).unwrap_or_else(|| "unknown".to_string());
        markets.push(MarketDescriptor {
            venue: Venue::Kalshi,
            market_id: ticker.clone(),
            title: optional_string(value, &["title", "subtitle", "yes_sub_title"]),
            slug: optional_string(value, &["event_ticker"]),
            status: status.clone(),
            book_enabled: optional_string(value, &["market_type"])
                .is_none_or(|kind| kind.eq_ignore_ascii_case("binary")),
            accepting_orders: status.eq_ignore_ascii_case("open")
                || status.eq_ignore_ascii_case("active"),
            close_time: optional_string(
                value,
                &["close_time", "expiration_time", "expected_expiration_time"],
            ),
            instruments: vec![OutcomeInstrument {
                outcome: None,
                instrument_id: ticker,
            }],
        });
    }
    Ok(MarketPage {
        venue: Venue::Kalshi,
        markets,
        next_cursor: optional_string(&root, &["cursor"]),
    })
}

pub fn parse_polymarket_market_info(text: &str) -> Result<Vec<OutcomeInstrument>, DiscoveryError> {
    let root: Value = serde_json::from_str(text).map_err(DiscoveryError::Json)?;
    let tokens = root
        .get("t")
        .or_else(|| root.get("tokens"))
        .and_then(Value::as_array)
        .ok_or(DiscoveryError::Field("tokens"))?;
    let instruments: Result<Vec<_>, _> = tokens
        .iter()
        .map(|token| {
            Ok(OutcomeInstrument {
                outcome: optional_string(token, &["o", "outcome"]),
                instrument_id: required_string(token, &["t", "token_id"])?,
            })
        })
        .collect();
    let instruments = instruments?;
    if instruments.is_empty() {
        return Err(DiscoveryError::Field("tokens"));
    }
    Ok(instruments)
}

async fn request_text(url: Url) -> Result<String, DiscoveryError> {
    let client = Client::builder()
        .timeout(Duration::from_secs(10))
        .user_agent("binary-market-data/0.1")
        .build()
        .map_err(DiscoveryError::Request)?;
    client
        .get(url)
        .send()
        .await
        .and_then(reqwest::Response::error_for_status)
        .map_err(DiscoveryError::Request)?
        .text()
        .await
        .map_err(DiscoveryError::Request)
}

fn endpoint(base: &str, relative: &str) -> Result<Url, DiscoveryError> {
    let base = format!("{}/", base.trim_end_matches('/'));
    let parsed = Url::parse(&base).map_err(|_| DiscoveryError::InvalidEndpoint)?;
    if !matches!(parsed.scheme(), "http" | "https") {
        return Err(DiscoveryError::InvalidEndpoint);
    }
    parsed
        .join(relative)
        .map_err(|_| DiscoveryError::InvalidEndpoint)
}

fn pair_instruments(
    token_ids: Vec<String>,
    outcomes: Vec<String>,
) -> Result<Vec<OutcomeInstrument>, DiscoveryError> {
    if token_ids.is_empty() {
        return Ok(Vec::new());
    }
    if !outcomes.is_empty() && outcomes.len() != token_ids.len() {
        return Err(DiscoveryError::Field("outcomes"));
    }
    Ok(token_ids
        .into_iter()
        .enumerate()
        .map(|(index, instrument_id)| OutcomeInstrument {
            outcome: outcomes.get(index).cloned(),
            instrument_id,
        })
        .collect())
}

fn string_list(value: Option<&Value>) -> Result<Vec<String>, DiscoveryError> {
    let Some(value) = value else {
        return Ok(Vec::new());
    };
    let parsed;
    let values = match value {
        Value::Array(values) => values,
        Value::String(text) => {
            parsed = serde_json::from_str::<Value>(text).map_err(DiscoveryError::Json)?;
            parsed
                .as_array()
                .ok_or(DiscoveryError::Field("string_array"))?
        }
        _ => return Err(DiscoveryError::Field("string_array")),
    };
    values
        .iter()
        .map(|value| {
            value
                .as_str()
                .filter(|text| !text.is_empty())
                .map(ToOwned::to_owned)
                .ok_or(DiscoveryError::Field("string_array.item"))
        })
        .collect()
}

fn required_string(value: &Value, keys: &[&str]) -> Result<String, DiscoveryError> {
    optional_string(value, keys).ok_or(DiscoveryError::Field("identifier"))
}

fn optional_string(value: &Value, keys: &[&str]) -> Option<String> {
    keys.iter().find_map(|key| {
        value
            .get(*key)
            .and_then(Value::as_str)
            .filter(|text| !text.is_empty())
            .map(ToOwned::to_owned)
    })
}

fn optional_bool(value: &Value, keys: &[&str]) -> Option<bool> {
    keys.iter()
        .find_map(|key| value.get(*key).and_then(Value::as_bool))
}

fn market_status(active: bool, closed: bool, archived: bool) -> String {
    if archived {
        "archived"
    } else if closed {
        "closed"
    } else if active {
        "active"
    } else {
        "inactive"
    }
    .to_string()
}

fn filter_page(page: &mut MarketPage, query: Option<&str>, limit: usize) {
    page.markets.retain(|market| market.book_enabled);
    if let Some(query) = query.map(str::trim).filter(|query| !query.is_empty()) {
        let query = query.to_ascii_lowercase();
        page.markets.retain(|market| {
            market.market_id.to_ascii_lowercase().contains(&query)
                || market
                    .title
                    .as_deref()
                    .is_some_and(|value| value.to_ascii_lowercase().contains(&query))
                || market
                    .slug
                    .as_deref()
                    .is_some_and(|value| value.to_ascii_lowercase().contains(&query))
        });
    }
    page.markets.truncate(limit);
}
